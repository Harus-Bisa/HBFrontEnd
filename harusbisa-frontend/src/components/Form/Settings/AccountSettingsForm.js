import React from "react";
import {connect} from "react-redux";
import { Button, Form, FormGroup, Label, Input, Col } from 'reactstrap';

function AccountSettingsForm(props){
    var [firstName, setFirstName] = React.useState(props.firstName);
    var [lastName, setLastName] = React.useState(props.lastName);
    var [school, setSchool] = React.useState(props.school);
    var [email, setEmail] = React.useState(props.email);

    var [oldPassword, setOldPassword] = React.useState();
    var [newPassword, setNewPassword] = React.useState();
    var [verifyNewPassword, setVerifyNewPassword] = React.useState();

    const submitProfile = (event) =>{
        event.preventDefault();
        props.changeProfile()
    }

    if(props.loading){
        return <p>Loading</p>
    }
    return(
        <div>
            <div className="content" style={{borderBottom:"2px solid #F4F4F4", paddingLeft: 0, paddingRight:0}}>
                <div>
                    <h5>Basic</h5>
                </div>
                <div className="row justify-content-end">
                    <div className="col-md-11">
                        <Form onSubmit={submitProfile} id="profile-form">
                            <FormGroup row>
                                <Label sm={5}>Nama Depan</Label>
                                <Col sm={7}>
                                    <Input type="text" id="firstName" value={firstName} onChange={(event) => setFirstName(firstName = event.target.value) }/>
                                </Col>
                            </FormGroup>
                            <FormGroup row>
                                <Label sm={5}>Nama Belakang</Label>
                                <Col sm={7}>
                                    <Input type="text" id="lastName" value={lastName} onChange={(event) => setLastName(lastName = event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <FormGroup row>
                                <Label sm={5}>Nama Universitas</Label>
                                <Col sm={7}>
                                    <Input type="text" id="school" value={school} onChange={(event) => setSchool(school= event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <FormGroup row>
                                <Label sm={5}>Email</Label>
                                <Col sm={7}>
                                    <Input type="email" id="email" value={email} onChange={(event) => setEmail(email = event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <div className="row justify-content-end">
                                <div className="col-sm-7">
                                    <Button type="submit" className="neutral-button" style={{width:'100%'}}>Simpan perubahan</Button>
                                </div>
                            </div>
                        </Form>
                    </div>
                </div>
            </div>
            
            <div className="content" style={{paddingLeft: 0, paddingRight:0}}>
                <h5>Password</h5>
                <div className="row justify-content-end">
                    <div className="col-md-11">
                        <p style={{marginBottom:'2rem'}}>Dengan membuat password memastkan anda dapat log in ke akun Harus Bisa anda</p>
                        <Form onSubmit={submitProfile} id="password-form">
                            <FormGroup row>
                                <Label sm={5}>Password Lama</Label>
                                <Col sm={7}>
                                    <Input type="text" id="oldPassword" value={oldPassword} onChange={(event) => setOldPassword(oldPassword= event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <FormGroup row>
                                <Label sm={5}>Password Baru</Label>
                                <Col sm={7}>
                                    <Input type="text" id="newPassword" value={newPassword} onChange={(event) => setNewPassword(newPassword= event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <FormGroup row>
                                <Label sm={5}>Ulangi Password Baru</Label>
                                <Col sm={7}>
                                    <Input type="text" id="verifyNewPassword" value={verifyNewPassword} onChange={(event) => setVerifyNewPassword(verifyNewPassword= event.target.value)}/>
                                </Col>
                            </FormGroup>
                            <div className="row justify-content-end">
                                <div className="col-sm-7">
                                    <Button type="submit" className="neutral-button" style={{width:'100%'}}>Simpan perubahan</Button>
                                </div>
                            </div>
                        </Form>
                    </div>
                </div>
            </div> 
        </div>
    )
}

function mapStateToProps(state){
    return{
        firstName: state.firstName,
        lastName: state.lastName,
        email: state.email,
        school: state.school,
        loading: state.loading
    }
}

AccountSettingsForm.defaultProps ={
    changeProfile: (() =>{})
}
export default connect(mapStateToProps)(AccountSettingsForm);