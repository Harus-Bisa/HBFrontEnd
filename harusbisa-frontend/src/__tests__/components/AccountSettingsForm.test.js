import React from "react";
import {mount} from "enzyme";
import configureMockStore from "redux-mock-store";
import '../../test-config';
import {Provider} from 'react-redux';
import AccountSettingsForm from "../../components/Form/Settings/AccountSettingsForm";
import { Form } from "reactstrap";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);
describe("AccountSettingsForm component", () =>{
    let initialState;
    let component;
    let store;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, setState])

    beforeEach(() =>{
        initialState = {
            loading: false,
            userId: "5db6093a7f75df3ce6c12095",
            firstName: "Wilson",
            lastName: "Burnawan",
            email: "wilson.burnawan@gmail.com",
            school: "University of Illinois at Urbana-Champaign",
            role: "faculty"
        }
        store = mockStore(initialState);
        store.dispatch = jest.fn()
        component = mount(<Provider store={store}><AccountSettingsForm/></Provider>)
    })
    
    afterEach(()=>{
        jest.clearAllMocks();
    })
    it("Matches snapshot", () =>{
        expect(component.html()).toMatchSnapshot();
    })
    it("Change state with form entry", () =>{
        const ids = ["firstName", "lastName", "school", "email", "oldPassword", "newPassword", "verifyNewPassword"];
        
        ids.forEach(id =>{
            var formEntry = component.find("#"+id).at(0)
            expect(setState).not.toHaveBeenCalled()
            formEntry.simulate("change")
            expect(setState).toHaveBeenCalledTimes(1)
            jest.clearAllMocks()
        })
    })
    it("Simulate submit", () =>{
        const mockPreventDefault = jest.fn();
        const formIds = ["profile-form","password-form"];

        formIds.forEach(id =>{
            var form = component.find("#"+id).at(0)
            expect(mockPreventDefault).not.toHaveBeenCalled()
            if (id === "password-form"){
                let passwordIds = ["oldPassword", "newPassword", "verifyNewPassword"]
                passwordIds.forEach(id =>{
                    var formEntry = component.find("#"+id).at(0)
                    formEntry.simulate("change",{target:{value:"pass"}})
                    component.update()
                })
            }
            form.simulate('submit',{preventDefault:mockPreventDefault, target:{id:id}})
            expect(store.dispatch).toHaveBeenCalled()
            expect(mockPreventDefault).toHaveBeenCalled()
            jest.clearAllMocks()
        })
    })
    it("Simulate delete", () =>{
       var deleteButton = component.find("#delete-button").at(0)
        expect(store.dispatch).not.toHaveBeenCalled();
        deleteButton.simulate("click")
        expect(store.dispatch).toHaveBeenCalled();
    })

    it("Simulate loading", () =>{
        initialState = {
            loading: true
        }
        store = mockStore(initialState);
        component = mount(<Provider store={store}><AccountSettingsForm /></Provider>)
        var loading = component.find(<p>Loading</p>)
        expect(loading).toBeDefined();
    })
})