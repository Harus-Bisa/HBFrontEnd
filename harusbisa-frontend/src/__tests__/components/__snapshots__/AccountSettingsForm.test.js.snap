// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSettingsForm component Matches snapshot 1`] = `"<div><div class=\\"content\\" style=\\"border-bottom: 2px solid #F4F4F4; padding-left: 0px; padding-right: 0px;\\"><div><h5>Basic</h5></div><div class=\\"row justify-content-end\\"><div class=\\"col-md-11\\"><form id=\\"profile-form\\" class=\\"\\"><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Nama Depan</label><div class=\\"col-sm-7\\"><input id=\\"firstName\\" type=\\"text\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Nama Belakang</label><div class=\\"col-sm-7\\"><input id=\\"lastName\\" type=\\"text\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Nama Universitas</label><div class=\\"col-sm-7\\"><input id=\\"school\\" type=\\"text\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Email</label><div class=\\"col-sm-7\\"><input id=\\"email\\" type=\\"email\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row justify-content-end\\"><div class=\\"col-sm-7\\"><button type=\\"submit\\" style=\\"width: 100%;\\" class=\\"neutral-button btn btn-secondary\\">Simpan perubahan</button></div></div></form></div></div></div><div class=\\"content\\" style=\\"padding-left: 0px; padding-right: 0px;\\"><h5>Password</h5><div class=\\"row justify-content-end\\"><div class=\\"col-md-11\\"><p style=\\"margin-bottom: 2rem;\\">Dengan membuat password memastkan anda dapat log in ke akun Harus Bisa anda</p><form id=\\"password-form\\" class=\\"\\"><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Password Lama</label><div class=\\"col-sm-7\\"><input id=\\"oldPassword\\" type=\\"password\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Password Baru</label><div class=\\"col-sm-7\\"><input id=\\"newPassword\\" type=\\"password\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-5 col-form-label\\">Ulangi Password Baru</label><div class=\\"col-sm-7\\"><input id=\\"verifyNewPassword\\" type=\\"password\\" class=\\"form-control\\" value=\\"\\"></div></div><div class=\\"row justify-content-end\\"><div class=\\"col-sm-7\\"><button type=\\"submit\\" style=\\"width: 100%;\\" class=\\"neutral-button btn btn-secondary\\">Simpan perubahan</button></div></div></form></div></div></div></div>"`;
